contents/blog/clickhouse-vs-postgres.md

```diff
@@ -0,0 +1,186 @@
+---
+date: 2022-10-05
+title: "Postgres versus ClickHouse, Grapes versus Grapefruits"
+rootPage: /blog
+sidebar: Blog
+showTitle: true
+hideAnchor: true
+author: ["mathew-pregasen"]
+featuredImage: ../images/blog/posthog-company-culture-blog.png
+featuredImageType: full
+categories: ["Engineering", "Guides"]
+---
+
+Honestly, it is a bit ridiculous to compare Postgres and ClickHouse. The two database solutions are as similar as grapes and grapefruit. ClickHouse was created for a specific purpose; Postgres was designed to be flexible and all-purpose. 
+
+So why even compare them? Because most companies that invest in an OLAP (online analytical processing) database like ClickHouse originally used an OLTP (online transaction processing) stack like mySQL or Postgres. PostHog’s database journey was no different. 
+
+In 2020, PostHog used Postgres to store client data. In the beginning, it worked. But usage grew very, very fast. Eventually, that all-purpose Postgres database was tasked to store millions of rows of data. At the rate data was being ingested, PostHog was on a collision course towards a major crash. 
+
+At first, the team tried a ton of hack-y and wacky solutions in attempts to get Postgres to work. Turns out, that wasn’t very sustainable (who would’ve thought!). Then, PostHog [migrated client data to ClickHouse](/blog/how-we-turned-clickhouse-into-our-eventmansion). Boom! 
+
+This is what that felt like...
+
+![thanos meme](../images/blog/clickhouse-vs-postgres/thanos-meme.png)  
+
+Suddenly, all those data problems were solved, Thanos-snapped from existence. Here, we’re diving deep into how and why. 
+
+## Databases 101: OLAP vs OLTP (aka columns vs rows) 
+
+If you’ve ever taken a databases 101 course, you’ve likely heard lectures on row-based relational databases. Good chance, the professor referred to them as simply *relational databases* or even *normal databases*. 
+
+The majority of popular solutions — mySQL, Postgres, SQLite — are all row-based. In each of these, data / objects are stored as rows, like a Yellowpages book. 
+
+In contrast, ClickHouse is a *columnar* database. Built by a single C++ binary, ClickHouse tables in memory are inverted — data is ingested as a column, meaning you’ve a *very large* number of columns and a much smaller set of rows.
+
+Here's what that looks like...
+
+![clickhouse vs postgres rows and columns](../images/blog/clickhouse-vs-postgres/rows-vs-columns.png)  
+
+This difference – to be clear – details how the data is *stored*; to the user, no mental-inversion is needed. You’re still dealing with tables with entries. You will continue to utilize SQL to interface with ClickHouse. The big difference is that those queries will perform very differently from analog queries to Postgres or any other row-based relational database. 
```

Comment:
I know this runs contrary to our use of US English, but I make an exception for British-style en dashes with a space vs em dashes without because... well, em dashes are just fugly! 😂

Reply:
Yeah I'm not a big fan either (I basically never use them), whatever you decide is fine with me.